src/ApiClient.ts

```diff
@@ -0,0 +1,204 @@
+  Cursor,
+  Envelope,
+  MessageApi,
+  PagingInfo,
+  PublishRequest,
+  QueryRequest,
+  SortDirection,
+  SubscribeRequest,
+} from '@xmtp/proto'
+import { retry } from './utils'
+import { NotifyStreamEntityArrival } from '@xmtp/proto/ts/dist/types/fetch.pb'
+
+const RETRY_SLEEP_TIME = 100
+
+export type QueryParams = {
+  startTime?: Date
+  endTime?: Date
+  contentTopics: string[]
+}
+
+export type QueryAllOptions = {
+  direction?: SortDirection
+  limit?: number
+}
+
+export type QueryStreamOptions = Omit<QueryAllOptions, 'limit'> & {
+  pageSize?: number
+}
+
+export type PublishParams = {
+  contentTopic: string
+  message: Uint8Array
+  timestamp?: Date
+}
+
+export type SubscribeParams = {
+  contentTopics: string[]
+}
+
+export type ApiClientOptions = {
+  maxRetries?: number
+}
+
+export type SubscribeCallback = NotifyStreamEntityArrival<Envelope>
+
+const toNanoString = (d: Date | undefined): undefined | string => {
+  return d && (d.valueOf() * 1_000_000).toFixed(0)
+}
+
+export default class ApiClient {
+  pathPrefix: string
+  maxRetries: number
+
+  constructor(pathPrefix: string, opts?: ApiClientOptions) {
+    this.pathPrefix = pathPrefix
+    this.maxRetries = opts?.maxRetries || 5
+  }
+
+  // Raw method for querying the API
+  private _query(req: QueryRequest): ReturnType<typeof MessageApi.Query> {
+    return retry(
+      MessageApi.Query,
+      [req, { pathPrefix: this.pathPrefix }],
+      this.maxRetries,
+      RETRY_SLEEP_TIME
+    )
+  }
+
+  // Raw method for publishing to the API
+  private _publish(req: PublishRequest): ReturnType<typeof MessageApi.Publish> {
+    return retry(
+      MessageApi.Publish,
+      [req, { pathPrefix: this.pathPrefix }],
+      this.maxRetries,
+      RETRY_SLEEP_TIME
+    )
+  }
+
+  // Raw method for subscribing
+  private _subscribe(
+    req: SubscribeRequest,
+    cb: NotifyStreamEntityArrival<Envelope>
+  ): ReturnType<typeof MessageApi.Subscribe> {
+    return retry(
+      MessageApi.Subscribe,
+      [req, cb, { pathPrefix: this.pathPrefix }],
+      this.maxRetries,
+      RETRY_SLEEP_TIME
+    )
+  }
+
+  async queryAll(
+    params: QueryParams,
+    {
+      direction = SortDirection.SORT_DIRECTION_ASCENDING,
+      limit,
+    }: QueryAllOptions
+  ): Promise<Envelope[]> {
+    const out: Envelope[] = []
+    // Use queryStreamPages for better performance. 1/100th the number of Promises to resolve compared to queryStream
+    for await (const page of this.queryStreamPages(params, {
+      direction,
+      // If there is a limit of < 100, use that as the page size. Otherwise use 100 and stop if/when limit reached.
+      pageSize: limit && limit < 100 ? limit : 100,
+    })) {
+      for (const envelope of page) {
+        out.push(envelope)
+        if (limit && out.length === limit) {
+          break
+        }
+      }
+    }
+    return out
+  }
+
+  // Will produce an AsyncGenerator of Envelopes
+  // Uses queryStreamPages under the hood
+  async *queryStream(
```

Comment:
Just for my own knowledge: Is `QueryStream` a standard term? Asking because at first glance I'd expect a "stream" to work similarly to `subscribe()`. Wondering if maybe this could just be `query()` and `queryPages()`?

Reply:
You know, maybe the term is confusing.

I could go for `query` and `queryIterator`?

Reply:
I renamed